hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/transfer/CustomFeeAssessmentStep.java

```diff
@@ -16,14 +16,261 @@
+     * It iterates through the token transfer list and assesses custom fees for each token transfer.
+     * It creates 2 new hashmaps with assessed hbar fees and assessed token fees.
+     * It is possible the assessed token custom fees could trigger custom fee again.
+     * So, we repeat the process for the assessed custom fees one more time.
+     *
+     * @param transferContext - transfer context
+     * @return - transaction body with assessed custom fees
+     */
+    public List<CryptoTransferTransactionBody> assessCustomFees(@NonNull final TransferContext transferContext) {
+        requireNonNull(transferContext);
+
+        final var handleContext = transferContext.getHandleContext();
+        final var tokenStore = handleContext.readableStore(ReadableTokenStore.class);
+        final var ledgerConfig = handleContext.configuration().getConfigData(LedgerConfig.class);
+        final var tokensConfig = handleContext.configuration().getConfigData(TokensConfig.class);
+        final var maxTransfersAllowed = ledgerConfig.xferBalanceChangesMaxLen();
+        final var maxCustomFeeDepth = tokensConfig.maxCustomFeeDepth();
+        final var recordBuilder = handleContext.recordBuilder(CryptoTransferRecordBuilder.class);
+        // list of total assessed custom fees to be added to the record
+        final List<AssessedCustomFee> customFeesAssessed = new ArrayList<>();
+        // the transaction to be assessed
+        var txnToAssess = op;
+        // list of assessed transactions, to be fed into further steps
+        final List<CryptoTransferTransactionBody> assessedTxns = new ArrayList<>();
+
+        // The first assessment inputs
+        var tokenTransfers = op.tokenTransfersOrElse(emptyList());
+        var hbarTransfers = op.transfersOrElse(TransferList.DEFAULT).accountAmountsOrElse(emptyList());
+
+        do {
+            validateTrue(levelNum <= maxCustomFeeDepth, CUSTOM_FEE_CHARGING_EXCEEDED_MAX_RECURSION_DEPTH);
+            // The result after each assessment
+            final var result = assessCustomFeesFrom(hbarTransfers, tokenTransfers, tokenStore, maxTransfersAllowed);
+
+            // when there are adjustments made to given transaction, need to re-build the transaction
+            final var modifiedInputBody = changedInputTxn(txnToAssess, result);
+            assessedTxns.add(modifiedInputBody);
+
+            validateTotalAdjustments(modifiedInputBody, maxTransfersAllowed);
+            customFeesAssessed.addAll(result.getAssessedCustomFees());
+
+            // build body from assessed custom fees to be fed to next level of assessment
+            txnToAssess = buildBodyFromAdjustments(result);
+
+            tokenTransfers = txnToAssess.tokenTransfersOrElse(emptyList());
+            hbarTransfers = txnToAssess.transfersOrElse(TransferList.DEFAULT).accountAmountsOrElse(emptyList());
+
+            levelNum++;
+        } while (!tokenTransfers.isEmpty());
```

Comment:
As is my custom when I see an unbounded loop 😛, I must insist we add a condition here like,
```
private static final int MAX_PLAUSIBLE_LEVEL_NUM = 10;
...
} while (!tokenTransfers.isEmpty() && levelNum <= MAX_PLAUSIBLE_LEVEL_NUM);

if (levelNum > MAX_PLAUSIBLE_LEVEL_NUM) {
    log.error("Recursive charging exceeded maximum plausible depth for transaction {}", op);
    throw new IllegalStateException("Custom fee charging overflow");
}
```

Reply:
But we are checking `levelNum` as first step in the loop. So it should fail after 2 levels right ?